kotlinx-coroutines-core/common/src/channels/ConflatedChannel.kt

```diff
@@ -120,12 +115,20 @@ internal open class ConflatedChannel<E> : AbstractChannel<E>() {
     }
 
     protected override fun onCancelIdempotent(wasClosed: Boolean) {
-        if (wasClosed) {
-            lock.withLock {
-                value = EMPTY
-            }
+        var elementCancelException: ElementCancelException? = null // resource cancel exception
+        lock.withLock {
+            elementCancelException = updateValueLocked(EMPTY)
         }
         super.onCancelIdempotent(wasClosed)
+        elementCancelException?.let { throw it } // throw exception at the end if there was one
+    }
+
+    private fun updateValueLocked(element: Any?): ElementCancelException? {
```

Comment:
Honestly, it's hard to imagine that someone creates a channel of resources that are being conflated and thrown away. Such solution is probably a bad design and should be avoided (+potentially blocking close() under the lock that blocks even concurrent `offer`s to the same channel!)

Let's prohibit non-null `onElementCancel` in conflated channels

Reply:
That would be weird from a composability standpoint, e.g. when a certain combo on `Channel(...)` constructor params is prohibited.  Too many things to document. Also, when we start supporting `onUndeliveredElement` in flows (in a similar way to reactive `onDiscard`) it would immediately show itself up. Let's discuss this separately.